kernel/ptrace.c

```diff
@@ -0,0 +1,243 @@
+#include "ptrace.h"
+#include "kernel/calls.h"
+#include "kernel/errno.h"
+#include "kernel/signal.h"
+#include "task.h"
+#include <string.h>
+
+// Returns stopped child with the given pid, locked with stop_lock
+static struct task *find_child(struct task *parent, pid_t_ pid) {
+    struct task *child = NULL;
+    list_for_each_entry(&parent->children, child, siblings) {
+        if (child->pid == pid) {
+            lock(&child->ptrace.stop_lock);
+            if (child->ptrace.stopped) {
+                break;
+            } else {
+                unlock(&child->ptrace.stop_lock);
+            }
+        }
+    }
+    return child;
+}
+
+dword_t sys_ptrace(dword_t request, dword_t pid, addr_t addr, dword_t data) {
+    switch (request) {
+    case PTRACE_TRACEME_:
+        STRACE("ptrace(PTRACE_TRACEME, %d, %#x, %#x)", pid, addr, data);
+        current->traced = true;
+        return 0;
+    case PTRACE_PEEKTEXT_:
+    case PTRACE_PEEKDATA_: {
+        STRACE("ptrace(PTRACE_PEEKDATA, %d, %#x, %#x)", pid, addr, data);
+        dword_t peek;
+        struct task *child = find_child(current, pid);
+        if (child) {
```

Comment:
Should be `if (child == NULL) return _EPERM;` to avoid having to indent the rest of the case (same with other cases)